Adding 'cooperative gestures' as an option when creating a map

ewagstaff opened the pull request:
## Launch Checklist

<!-- Thanks for the PR! Feel free to add or remove items from the checklist. -->


 - [x] Confirm **your changes do not include backports from Mapbox projects** (unless with compliant license) - if you are not sure about this, please ask!
 
 This does use the same option name as Mapbox uses, but `collaborativeGestures` seems to be the standard way to describe this behavior. All of the code was written from scratch.
 
 - [x] Briefly describe the changes in this PR.
 
 The main additions are in `map.ts` where I added the option for collaborative gestures when the map is created. If the boolean is true, this update will create a hidden div containing gesture instructions that gets shown and hidden by adding and removing a CSS class.
 
 Small additions were made in `scroll_zoom.ts` and `touch_pan.ts` to add the classes to show the div when appropriate.
 
 - [x] Include before/after visuals or gifs if this PR includes visual changes.

With the option off:
![coopoff](https://user-images.githubusercontent.com/2423322/168396395-5022f34e-d38e-438e-8850-498661d9c611.gif)

With the option on:
![coop](https://user-images.githubusercontent.com/2423322/168396424-b0f51f2a-988c-4783-99ea-a2e9f89fb400.gif)

 - [ ] Write tests for all new functionality.
 I've asked in the #maplibre OSM channel for guidance on how to structure a test of this new option. If a maintainer could point me in the direction of a similar example, a reference would be really helpful.
 
 - [x] Document any changes to public APIs.
 - [ ] Post benchmark scores.
Maybe my computer is not dialed enough to run these benchmarks, because when I run `npm run benchmarks` it fires up the processes but hasn't completed after a few hours.
 
 - [x] Manually test the debug page.
 - [x] Suggest a changelog category: Feature
 - [x] Add an entry inside this element for inclusion in the `maplibre-gl-js` changelog: `<changelog>Added collaborativeGestures option when instantiating map to prevent inadvertent scrolling/panning when navigating a page where map is embedded inline #234 </changelog>`.

HarelM commented on src/ui/map.ts:
```diff
@@ -2384,6 +2394,28 @@ class Map extends Camera {
+        const container = this._container;
+        this._metaPress = false;
+        const cooperativeGestureScreen = this._cooperativeGesturesScreen = DOM.create('div', 'maplibregl-cooperative-gesture-screen', container);
+        let modifierKeyPrefix = "^"; // control key
```
I'm not sure all the users know ^ stands for control key.
I would suggest to write the word "Ctrl"

ewagstaff replied:
Great point, will update.

HarelM commented on src/ui/map.ts:
```diff
@@ -2439,6 +2471,17 @@ class Map extends Camera {
+            this._cooperativeGesturesScreen.classList.add("show")
+            setTimeout(() => {
+                this._cooperativeGesturesScreen.classList.remove("show")
+            }, 100)
```
100 seems like a very little time, is it more because of the animation?

ewagstaff replied:
Yep, 100 ms is just to trigger the CSS-based transition, the full animation is 1 second at full opacity with a 1 second fade after.

HarelM commented:
Thanks for taking the time to open this PR!
What would happen for pinch-zoom and two finger pitch change if this is set to true?
Same question for control + mouse drag to change pitch on desktop?

Did you test this on a map that is not full screen and made sure it behaves as expected as this is probably the main use case for this feature?

HarelM commented on src/css/maplibre-gl.css:
```diff
@@ -917,3 +917,39 @@ a.mapboxgl-ctrl-logo.mapboxgl-compact {
+}
+
+@media (max-width: 767px) {
+    .maplibregl-cooperative-gesture-screen .desktop-message {
```
Is this the right way to test of the map is used in mobile or desktop? What about tablets?

ewagstaff replied:
Thanks for the question, and looking closely at this media query around hover feature detection that [has pretty good support](https://caniuse.com/css-media-interaction), I think `@media (hover: none)` will be the more reliable check that will also work on large tablets.

ewagstaff commented:
> Thanks for taking the time to open this PR! What would happen for pinch-zoom and two finger pitch change if this is set to true? Same question for control + mouse drag to change pitch on desktop?
> 
> Did you test this on a map that is not full screen and made sure it behaves as expected as this is probably the main use case for this feature?

Pinch-zoom and two-finger pitch change would be unaffected, because they would both successfully pass the requirement of 2 minimum touches to bypass the instruction screen.

Yep, here's a quick gif of an inline that that demonstrates how the new feature enables navigation without getting stuck on the map:
![ezgif-4-9fc55ac5f2](https://user-images.githubusercontent.com/2423322/168697755-173b72a1-8c5a-42c2-8421-b30412448ccc.gif)

HarelM commented:
Can you share a stackblitz with this version so I can try it out?
From my understanding panning the map from bottom to upper part should be problematic on mobile, but I need to test it...

HarelM commented:
I suggest to run
npm run lint -- --fix 
before commiting to fix easy lint errors.

ewagstaff commented:
Lint commands are returning without errors and I have a demo with the build files posted here: https://js-ewvhew.stackblitz.io

wipfli commented:
Thanks at @ewagstaff. I tried the demo on mobile and instead of panning, the map pitched when I used two fingers.

ewagstaff commented:
> Thanks at @ewagstaff. I tried the demo on mobile and instead of panning, the map pitched when I used two fingers.

Ah @wipfli I do see what you mean. Typically I have that gesture disabled. Is there a difference between what you'd expect to do for pitch change vs panning vertically in terms of how your fingers would move?

wipfli commented:
I also usually turn pitch off. The gestures for pitch and pan vertically are the same for me.

wipfli commented:
I wouldn't enforce something like a mutex. But maybe add a comment in the description

wipfli commented:
I would let the user decide the settings freely. We can help them by telling about the pitch stuff in the description of the feature. Would also be great to add an example to the docs website and mention pitch there.

HarelM commented:
Even if we let the user decide we still have a feature that contradicts another feature.
If someone would like to allow pitch and this feature there's no way for this to work correctly.
This is a problem from my point of view that needs to be fixed or addressed before this can be merged.

ewagstaff commented:
Looking at other examples deployed live, both Google and Mapbox use a three-finger pitch gesture when cooperative gestures is enabled. What are your thoughts about that adjustment?

wipfli commented:
Sounds like a good solution to me...

ewagstaff commented:
Updated the PR with logic that will make pitch change only register with 3 fingers if coop gestures is enabled. The source on the stackblitz has been updated so you can demo: https://js-ewvhew.stackblitz.io/

HarelM commented:
Works a lot better! Thanks!
I've surfed from a mobile device and it should say something about 2-3 finger touch and not ctrl+ scroll.
I think this is the only comment I have.
In the docs it should be clear that enabling this changes a lot of behaviors related to touch.

acalcutt commented:
On my windows laptop + touchpad I wasn't able to figure out how to zoom in with this. two fingers on the touchpad scrolled the page like expected, but ctrl + two fingers zoomed the whole page (text at the top/bottom) and 3 fingers + ctrl either didn't do anything or brought up windows task switcher.

HarelM commented:
Right, that also a valid use case that needs to be taken into consideration.
@acalcutt can you check how Google maps behaves for this user case?

acalcutt commented:
I tested here ( https://developers.google.com/maps/documentation/javascript/examples/interaction-cooperative ) and the Ctrl + 2 fingers on the touchpad zoom into the map, instead of zooming into the page like happened with the test page above

ewagstaff commented:
@acalcutt thanks for catching this, please try the demo site again when you can: https://js-ewvhew.stackblitz.io/

HarelM commented:
This seems to work good in general I think. All controls are working in desktop and mobile (need 3 fingers to change pitch which is acceptable I guess).
The only problem I see is that when I surf to this page from a mobile device (Android Chrome) I get the "use control + scroll" to zoom the map instead of use 2 fingers to pan the map...
Other than that, changelog entry is needed and tests are needed to cover these scenarios.

ewagstaff commented:
@HarelM what device are you seeing that on? I see the "two fingers" language both on my iPhone 12 using Safari and an Android Galaxy Tab A using Chrome. It should swap out the language based on `hover` feature detection, and that should be [supported pretty broadly](https://caniuse.com/css-media-interaction).

For the changelog, I suggested this language: `<changelog>Added collaborativeGestures option when instantiating map to prevent inadvertent scrolling/panning when navigating a page where map is embedded inline #234 </changelog>`

To add that, would I just update the CHANGELOG.md file in the repo? Should I increment the version to 2.1.9?

Working on adding test cases now.

HarelM commented:
No need to increment version.
Yes, simply edit the top most entries in the change log, those without the version.
I have an Android 12, chrome, Samasung Galaxy S20.

ewagstaff commented:
I've added `cooperative_gestures.test.ts` in the src/ui folder, and it's passing when I run it individually. However, I'm unfamiliar with Jest -- does the file's presence in that directory include it in the pre-build test logic, or does it need to be explicitly added somewhere?

I've updated the changelog, and I added an additional media query check to display the "2 finger" language on any screen smaller than 480px if the feature media query fails. The demo should work on your S20 as expected: https://js-ewvhew.stackblitz.io/

HarelM commented on src/ui/map.ts:
```diff
@@ -2384,6 +2395,34 @@ class Map extends Camera {
+        }
+        this._cooperativeGesturesScreen.innerHTML = `
+            <div class="maplibregl-desktop-message">Use ${modifierKeyPrefix} + scroll to zoom the map</div>
+            <div class="maplibregl-mobile-message">Use two fingers to move the map</div>
```
Is there a way to localize this sentence? i.e. what happens if I want to translate this to a different language? If there isn't a straight forwards way, I think it would be good to add this as part of the documentation somehow... @wipfli  what do you think?

ewagstaff replied:
I can't think of a simple way to handle every language by default, but it might be possible for me to accept either `true` as the value of the `cooperativeGestures` flag or a config object that includes custom desktop / mobile instruction text. So it could be written in any language. Potentially helpful?

HarelM replied:
I would expand the object and add these values as default. Meaning in the object is null or undefined don't do this gesture, but if not use its fields for the relevant text.

HarelM replied:
I.e. Don't define an object that is either boolean or object, just an object that can be initialized easily with default values.
I think...

HarelM replied:
Since you are already using a dictionary, how about just having 3 fields are remove the complexity from the code itself:
```js
cooperativeGestures: {
  windowsHelpText: "Use Ctrl + scroll to zoom the map",
  macHelpText: "Use ⌘ + scroll to zoom the map",
  mobileHelpText: "Use two fingers to move the map",
  RTL: false //no sure this is needed, but it might be...?
}
```
I've added some RTL support, but I'm not 100% sure it's needed as I think one can address this using regular css matching, right?

ewagstaff replied:
Yep, if you wanted to restyle any of the elements, they should be easy to query and do so. I can go ahead and implement those 3 properties the way you've outlined above.

HarelM replied:
This world be great, thanks!

ewagstaff replied:
Updated to accept those 3 if they exist, fall back to defaults if they don't, and updated param docs to include an example.

HarelM commented:
Yes, tests that are in the src folder are automatically picked up in the unit test run.
I've tested now with my android mobile, windows and mac desktop and they all work as expected. 
Good job @ewagstaff!!

HarelM commented on src/ui/map.ts:
```diff
@@ -218,7 +218,13 @@ const defaultOptions = {
  * @param {boolean|Object} [options.touchZoomRotate=true] If `true`, the "pinch to rotate and zoom" interaction is enabled. An `Object` value is passed as options to {@link TouchZoomRotateHandler#enable}.
  * @param {boolean|Object} [options.touchPitch=true] If `true`, the "drag to pitch" interaction is enabled. An `Object` value is passed as options to {@link TouchPitchHandler#enable}.
- * @param {boolean} [options.cooperativeGestures=false] If `true`, map is only accessible on desktop while holding Command/Ctrl and only accessible on mobile with two fingers. Interacting with the map using normal gestures will trigger an informational screen. With this option enabled, "drag to pitch" requires a three-finger gesture.
+ * @param {boolean} [options.cooperativeGestures=false] If `true` or set to an options object, map is only accessible on desktop while holding Command/Ctrl and only accessible on mobile with two fingers. Interacting with the map using normal gestures will trigger an informational screen. With this option enabled, "drag to pitch" requires a three-finger gesture.
```
This is no longer a boolean, right?

ewagstaff replied:
It's checking for truthy values in most places, so `true` would still work to enable the option with the defaults. Should it follow the `boolean|Object` pattern for touchPitch above?

HarelM replied:
Boolean or ibject is a possibility, but an interface that has an object in it is a very bad interface as the user won't know what can be passed.
I prefer `boolean|GestureOptions` and define GestureOptions type some where, or you can define GestureOption type as boolean or the object with all the text etc, I.e 
```
export type GestureOptions = boolean | {
   macText: ...
}
```

HarelM commented on src/ui/map.ts:
```diff
@@ -273,7 +279,7 @@ class Map extends Camera {
     _controlPositions: {[_: string]: HTMLElement};
     _interactive: boolean;
-    _cooperativeGestures: boolean;
+    _cooperativeGestures: any;
```
Define a type, don't use `any`

HarelM replied:
See my comment above about the docs and the relevant type definition. It should be the same.

HarelM commented on src/ui/map.ts:
```diff
@@ -145,6 +146,7 @@ const defaultOptions = {
     doubleClickZoom: true,
     touchZoomRotate: true,
     touchPitch: true,
+    cooperativeGestures: false,
```
Undefined instead of false?

HarelM commented on src/ui/map.ts:
```diff
@@ -85,6 +85,7 @@ export type MapOptions = {
     doubleClickZoom?: boolean;
     touchZoomRotate?: boolean;
     touchPitch?: boolean;
+    cooperativeGestures?: boolean;
```
Should be the same type of the map's _cooperativeGesture

ewagstaff commented:
Per your suggestion @HarelM I added the `GestureOptions` type and used union types when defining the property. Let me know if that looks correct.
